File: fm-domain/src/main/java/io/zeitwert/fm/task/model/impl/AggregateWithTasksMixin.java
package io.zeitwert.fm.task.model.impl;

import io.dddrive.core.ddd.model.Aggregate;
import io.dddrive.core.obj.model.Obj;
import io.zeitwert.fm.task.model.DocTask;
import io.zeitwert.fm.task.model.DocTaskRepository;
import io.zeitwert.fm.task.model.ItemWithTasks;

import java.time.OffsetDateTime;
import java.util.List;

public interface AggregateWithTasksMixin extends ItemWithTasks {

	Aggregate aggregate();

	DocTaskRepository taskRepository();

	@Override
	default List<DocTask> getTasks() {
		Object id = this.aggregate().getId();
		String fkName = (this instanceof Obj) ? "relatedObjId" : "relatedDocId";
		return this.taskRepository().getByForeignKey(fkName, id).stream().map(it -> taskRepository().get(it)).toList();
	}

	@Override
	default DocTask addTask() {
		DocTask task = this.taskRepository().create(this.aggregate().getTenantId(), null, OffsetDateTime.now());
		task.setRelatedToId((Integer) this.aggregate().getId());
		return task;
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/task/model/impl/AggregateWithTasksMixin.kt
package io.zeitwert.fm.task.model.impl

import dddrive.app.doc.model.Doc
import dddrive.app.obj.model.Obj
import dddrive.ddd.model.Aggregate
import dddrive.query.query
import io.zeitwert.fm.task.model.DocTask
import io.zeitwert.fm.task.model.DocTaskRepository
import io.zeitwert.fm.task.model.ItemWithTasks
import java.time.OffsetDateTime

interface AggregateWithTasksMixin : ItemWithTasks {

	fun aggregate(): Aggregate

	fun taskRepository(): DocTaskRepository

	override val tasks: List<DocTask>
		get() {
			val id = this.aggregate().id
			val fkName = if (this is Obj) "relatedObjId" else "relatedDocId"
			val querySpec = query {
				filter { fkName eq id }
			}
			return this.taskRepository().find(querySpec).map { taskRepository().get(it) }
		}

	override fun addTask(
		userId: Any,
		timestamp: OffsetDateTime,
	): DocTask {
		val aggregate = aggregate()
		if (aggregate is Obj) aggregate.tenantId else (aggregate as Doc).tenantId
		val task = this.taskRepository().create()
		task.relatedToId = aggregate.id
		return task
	}

}
